### PR TITLE
New resource & data source 'azurerm_azuread_group'

### PR DESCRIPTION
**Pull request by tiwood**
This replaces PR #1585.

This PR adds a new resource and data source for Azure Active Directory groups.

- [x] New data source 'azurerm_azuread_group'
- [x] New resource 'azurerm_azuread_group'
- [x] Add markdown documentation
- [x] Add tests

## Example Usage

```hcl

// This creates a new Azure AD group with the display name "my_group"
resource "azurerm_azuread_group" "my_group" {
 name = "my_group"
}

// This queries an Azure AD group by ID
data "azurerm_azuread_group" "my_group_by_id" {
  object_id = "78722cfc-8946-11e8-95f1-2200ec79ad01"
}

// This queries an Azure AD group by Name
data "azurerm_azuread_group" "my_group_by_name" {
  name = "my_group"
}
```

The following attributes are exported:

* `id` - The Object ID for the Azure AD Group.
* `object_id` - The Object ID for the Azure AD Group.
* `name` - The Display Name for the Azure AD Group.

### CODE REVIEW

**Comment by tiwood** [comment]
Related: #1272

**Comment by tiwood** [comment]
@jeffreyCline, @tombuildsstuff, any plans to get this merged into 1.15?

File: azurerm/resource_arm_azuread_group.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	return &schema.Resource{
+		Create: resourceArmActiveDirectoryGroupCreate,
+		Read:   resourceArmActiveDirectoryGroupRead,
+		Delete: resourceArmActiveDirectoryGroupDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"object_id": {
```
Comment (katbyte) [inline azurerm/resource_arm_azuread_group.go]: Is there a reason for this property to exist? `.id` can be used instead and is even what your using in your tests ` object_id = "${azurerm_azuread_group.test.id}"`

File: website/docs/d/azuread_group.html.markdown [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_azuread_group"
+sidebar_current: "docs-azurerm-datasource-azuread-group"
+description: |-
+  Gets information about a group object within the Azure Active Directory.
```
Comment (katbyte) [inline website/docs/d/azuread_group.html.markdown]: This might read better as:
```
Gets information about an Azure Active Directory group.
```

File: website/docs/d/azuread_group.html.markdown [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+---
+
+# Data Source: azurerm_azuread_group
+
+Gets information about a Group object within the Azure Active Directory.
```
Comment (katbyte) [inline website/docs/d/azuread_group.html.markdown]: as commented about above

File: website/docs/d/azuread_group.html.markdown [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `object_id` - (Optional) The ID of the Azure AD Group we want to lookup.
+
+* `name` - (Optional) The ID of the Azure AD Group we want to loopup.
+
+-> **NOTE:** At least one of `name` or `object_id` must be specified.
```
Comment (katbyte) [inline website/docs/d/azuread_group.html.markdown]: This should be enforced by a `CustomizeDiff`

File: azurerm/resource_arm_azuread_group_test.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMActiveDirectoryGroupDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMActiveDirectoryGroupExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "name", fmt.Sprintf("acctest%s", id)),
+				),
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_azuread_group_test.go]: Could we get an import test step here?
```go
{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```

File: azurerm/resource_arm_azuread_group_test.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
```
Comment (katbyte) [inline azurerm/resource_arm_azuread_group_test.go]: Could we get an import test step here?
```go
{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```

File: azurerm/data_source_azuread_group.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+				Computed:      true,
+				ConflictsWith: []string{"name"},
+			},
+
+			"name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"object_id"},
+			},
```
Comment (katbyte) [inline azurerm/data_source_azuread_group.go]: Could we get some validation on this value?

File: azurerm/provider.go [hunk 8]
```diff
@@ -80,6 +80,7 @@ func Provider() terraform.ResourceProvider {
 		DataSourcesMap: map[string]*schema.Resource{
 			"azurerm_azuread_application":                   dataSourceArmAzureADApplication(),
 			"azurerm_azuread_service_principal":             dataSourceArmActiveDirectoryServicePrincipal(),
+			"azurerm_azuread_group":                         dataSourceArmAzureADGroup(),
```
Comment (katbyte) [inline azurerm/provider.go]: Could we sort these alphabetically?

File: azurerm/provider.go [hunk 9]
```diff
@@ -126,6 +127,7 @@ func Provider() terraform.ResourceProvider {
 			"azurerm_azuread_application":                     resourceArmActiveDirectoryApplication(),
 			"azurerm_azuread_service_principal":               resourceArmActiveDirectoryServicePrincipal(),
 			"azurerm_azuread_service_principal_password":      resourceArmActiveDirectoryServicePrincipalPassword(),
+			"azurerm_azuread_group":                           resourceArmActiveDirectoryGroup(),
```
Comment (katbyte) [inline azurerm/provider.go]: Could we sort these alphabetically?

**Comment by X-Guardian** [comment]
Hi @tiwood, are you able to make the requested changes to this PR? I would love to be able to use this data source instead of having to hard-code AD Group Ids in my Azure Terraform configurations.

**Comment by tiwood** [comment]
@X-Guardian, I'm going to make the required changes in the next few days.

**Comment by X-Guardian** [comment]
Thanks @tiwood , that's excellent news!

File: azurerm/resource_arm_azuread_group.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+func resourceArmActiveDirectoryGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).groupsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+
+	properties := graphrbac.GroupCreateParameters{
+		DisplayName:     &name,
+		MailEnabled:     utils.Bool(false),
```
[the remark this answers is not in the document]
Comment (tiwood) [reply azurerm/resource_arm_azuread_group.go]: I could do that, but at the moment the Graph API only supports the creation of security enabled groups (MailEnabled: false, SecurityEnabled: true). See [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/e231dc2f050b623f6e7e7b1e3f5d1d31fd88f239/vendor/github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac/models.go#L1937).

So what is the protocol here, adding these to the schema as computed and default to them for now, or keep it as is?

File: azurerm/resource_arm_azuread_group.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+			"object_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
[the remark this answers is not in the document]
Comment (tiwood) [reply azurerm/resource_arm_azuread_group.go]: I will do a NoZeroValues validation, as I'm unable to identity limitations on the display name at the moment.

File: azurerm/data_source_azuread_group.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"object_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"name"},
+			},
+
+			"name": {
```
[the remark this answers is not in the document]
Comment (tiwood) [reply azurerm/data_source_azuread_group.go]: Will be part of the next commit.

File: azurerm/data_source_azuread_group.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
+package azurerm
+
+import (
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
```
[the remark this answers is not in the document]
Comment (tiwood) [reply azurerm/data_source_azuread_group.go]: Will be part of the next commit.

File: azurerm/resource_arm_azuread_group_test.go (hunk 5 above)
Comment (tiwood) [reply azurerm/resource_arm_azuread_group_test.go]: Added as requested.

File: azurerm/resource_arm_azuread_group_test.go (hunk 6 above)
Comment (tiwood) [reply azurerm/resource_arm_azuread_group_test.go]: Added as requested.

**Comment by tiwood** [comment]
Hi @katbyte, I think this is ready for another review, please let me know if further changes are required.

**Comment by Leon99** [comment]
Huge thanks to @tiwood for contributing this. @katbyte any ideas if this can be reviewed/merged any time soon?

**Comment by tiwood** [comment]
maybe @tombuildsstuff can help reviewing this since @katbyte removed her assignment?

**Comment by tombuildsstuff** [comment]
hey @tiwood 

Thanks for this PR - apologies for the delayed review of this!

Just to give some context behind the delayed review of this: as a part of the upcoming v2.0 release of the AzureRM Provider we're thinking through some of the longer-term changes that we'd like to make to the Provider.

At this point we're trying to determine where the Azure Active Directory resources belong, since there's now enough edge-cases/feature enhancements that this may make sense in it's own Provider; which would allow us to more easily ship support for the AzureAD resources which are harder to test.

The proposal to split the AzureAD resources out into its own provider is being tracked in #2322 - would you be able to take a look and let us know your thoughts?

Thanks!

**Comment by tombuildsstuff** [comment]
👋 

As mentioned in #2322 - we're planning on splitting the AzureAD resources out into their own provider. Since this PR is waiting on that change (which is planned in the not-too-distant future) rather than leaving this open in a holding pattern I'm going to temporarily close this PR for the moment - however we should be able to migrate this PR over once the split-out's done (I hope you don't mind @tiwood).

Thanks!
